### PR TITLE
add RouterHpsmHlpBalancer

### PR DESCRIPTION
**Pull request by Preston-Harrison**

### CODE REVIEW

File: contracts/RouterHpsmHlpBalancer.sol [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+            tokenOut,
+            poolId,
+            hpsmAmountOut,
+            minOut,
+            _self,
+            signedQuoteData
+        );
+
+        uint256 amountOut = _balanceOfSelf(tokenOut);
+        require(amountOut >= minOut, "Insufficient amount out");
```
Comment (bitcoinbrisbane) [inline contracts/RouterHpsmHlpBalancer.sol:66]: I think this should be at LN45 to save gas and be safer

File: contracts/RouterHpsmHlpBalancer.sol [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+            amountIn,
+            minOut,
+            _self,
+            signedQuoteData
+        );
+
+        _hpsmWithdraw(fxToken, tokenOut, _balanceOfSelf(fxToken));
+
+        uint256 amountOut = _balanceOfSelf(tokenOut);
+        require(amountOut >= minOut, "Insufficient amount out");
```
Comment (bitcoinbrisbane) [inline contracts/RouterHpsmHlpBalancer.sol:117]: Same here

**Review by bitcoinbrisbane — COMMENTED** [review]
2 small things to check

File: contracts/RouterHpsmHlpBalancer.sol (hunk 1 above)
Comment (Preston-Harrison) [reply contracts/RouterHpsmHlpBalancer.sol:66]: The amountOut is not known until after the swaps have occurred, so it can only be checked at the end

**Review by bitcoinbrisbane — DISMISSED** [review]

File: tests/unit/RouterHpsmHlpBalancer.spec.ts [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  PSM,
+  ROUTER,
+  USDC
+} from "../addresses";
+import { fetchEncodedSignedQuotes } from "handle-sdk/dist/src/utils/h2so-utils";
+import { RouterHlpBalancer__factory } from "../../build/typechain/factories/RouterHlpBalancer__factory";
+
+chai.use(solidity);
+
+const ONE_HUNDRED_E18 = ethers.utils.parseUnits("100", 18);
```
Comment (rellfy) [inline tests/unit/RouterHpsmHlpBalancer.spec.ts:30]: Is this used anywhere?

**Review by nabsta-eth — DISMISSED** [review]

**Review by bazmatic — DISMISSED** [review]

**Review by bazmatic — APPROVED** [review]

**Review by rellfy — APPROVED** [review]

**Pull request merged**